packages/visx-text/src/hooks/useText.ts

```diff
@@ -0,0 +1,123 @@
+import { useMemo } from 'react';
+import reduceCSSCalc from 'reduce-css-calc';
+import { TextProps, WordsWithWidth } from '../types';
+import getStringWidth from '../util/getStringWidth';
+
+function isNumber(val: unknown): val is number {
+  return typeof val === 'number';
+}
+
+function isValidXOrY(xOrY: string | number | undefined) {
+  return (
+    // number that is not NaN or Infinity
+    (typeof xOrY === 'number' && Number.isFinite(xOrY)) ||
+    // for percentage
+    typeof xOrY === 'string'
+  );
+}
+
+export default function useText(
+  props: TextProps,
+): {
+  wordsByLines: WordsWithWidth[];
+  startDy: string;
+  transform: string;
+} {
+  const {
+    verticalAnchor = 'end',
+    scaleToFit = false,
+    angle,
+    width,
+    lineHeight = '1em',
+    capHeight = '0.71em', // Magic number from d3
+    children,
+    style,
+    ...textProps
+  } = props;
+
+  const { x = 0, y = 0 } = textProps;
+  const isXOrYValid = !isValidXOrY(x) || !isValidXOrY(y);
+
+  const { wordsWithWidth, spaceWidth } = useMemo(() => {
+    const words: string[] = children == null ? [] : children.toString().split(/(?:(?!\u00A0+)\s+)/);
+    return {
+      wordsWithWidth: words.map(word => ({
+        word,
+        wordWidth: getStringWidth(word, style) || 0,
+      })),
+      spaceWidth: getStringWidth('\u00A0', style) || 0,
+    };
+  }, [children, style]);
+
+  const wordsByLines = useMemo(() => {
+    if (isXOrYValid) {
+      return [];
+    }
+
+    // Only perform calculations if using features that require them (multiline, scaleToFit)
+    if (width || scaleToFit) {
+      return wordsWithWidth.reduce((result: WordsWithWidth[], { word, wordWidth }) => {
+        const currentLine = result[result.length - 1];
+
+        if (
+          currentLine &&
+          (width == null || scaleToFit || (currentLine.width || 0) + wordWidth + spaceWidth < width)
+        ) {
+          // Word can be added to an existing line
+          currentLine.words.push(word);
+          currentLine.width = currentLine.width || 0;
+          currentLine.width += wordWidth + spaceWidth;
+        } else {
+          // Add first word to line or word is too long to scaleToFit on existing line
+          const newLine = { words: [word], width: wordWidth };
+          result.push(newLine);
+        }
+
+        return result;
+      }, []);
+    }
+
+    return [
+      {
+        words: children == null ? [] : children.toString().split(/(?:(?!\u00A0+)\s+)/),
+      },
+    ];
+  }, [isXOrYValid, width, scaleToFit, children, wordsWithWidth, spaceWidth]);
+
+  const startDy = useMemo(() => {
+    const startDyStr = isXOrYValid
+      ? ''
```

Comment:
it looks like previously this case would result in `startyDy` being `undefined`, empty string is probably okay tho